src/rgw/rgw_crypt.cc

```diff
@@ -0,0 +1,1307 @@
+    }
+  }
+  if (bl.length() == 0) {
+    if (cache.length() > 0) {
+      /*flush cached data*/
+      bufferlist data;
+      crypt->encrypt(cache, 0, cache.length(), data, ofs);
+      res = next->handle_data(data, ofs, phandle, pobj, again);
+      ofs += cache.length();
+      cache.clear();
+      if (res != 0)
+        return res;
+    }
+    /*replicate 0-sized handle_data*/
+    res = next->handle_data(cache, ofs, phandle, pobj, again);
+  }
+  return res;
+}
+
+int RGWPutObj_BlockEncrypt::throttle_data(void *handle, const rgw_obj& obj, bool need_to_wait) {
+  return next->throttle_data(handle, obj, need_to_wait);
+}
+
+std::string create_random_key_selector() {
+  char random[AES_256_KEYSIZE];
+  if (get_random_bytes(&random[0], sizeof(random)) != 0) {
+    dout(0) << "ERROR: cannot get_random_bytes. " << dendl;
+    for (char& v:random) v=rand();
+  }
+  return std::string(random, sizeof(random));
+}
+
+int get_barbican_url(CephContext * const cct,
+                     std::string& url)
+{
+  url = cct->_conf->rgw_barbican_url;
+  if (url.empty()) {
+    ldout(cct, 0) << "ERROR: conf rgw_barbican_url is not set" << dendl;
+    return -EINVAL;
+  }
+
+  if (url[url.size() - 1] != '/') {
+    url.append("/");
+  }
+
+  return 0;
+}
+
+int request_key_from_barbican(CephContext *cct,
+                              boost::string_ref key_id,
+                              boost::string_ref key_selector,
+                              const std::string& barbican_token,
+                              std::string& actual_key) {
+  std::string secret_url;
+  if (get_barbican_url(cct, secret_url) < 0) {
+     return -EINVAL;
+  }
+  secret_url += "v1/secrets/" + std::string(key_id);
+
+  int res;
+  bufferlist secret_bl;
+  RGWHTTPTransceiver secret_req(cct, &secret_bl);
+  secret_req.append_header("Accept", "application/octet-stream");
+  secret_req.append_header("X-Auth-Token", barbican_token);
+
+  res = secret_req.process("GET", secret_url.c_str());
+  if (res < 0) {
+    return res;
+  }
+  if (secret_req.get_http_status() ==
+      RGWHTTPTransceiver::HTTP_STATUS_UNAUTHORIZED) {
+    return -EACCES;
+  }
+
+  if (secret_req.get_http_status() >=200 &&
+      secret_req.get_http_status() < 300 &&
+      secret_bl.length() == AES_256_KEYSIZE) {
+    actual_key = std::string(secret_bl.c_str(), secret_bl.length());
+    } else {
+      res = -EACCES;
+    }
+  return res;
+}
+
+int get_actual_key_from_kms(CephContext *cct,
+                            boost::string_ref key_id,
+                            boost::string_ref key_selector,
+                            std::string& actual_key)
+{
+  int res = 0;
+  ldout(cct, 20) << "Getting KMS encryption key for key=" << key_id << dendl;
+  if (key_id.starts_with("testkey-")) {
+    /* test keys for testing purposes */
+    boost::string_ref key = key_id.substr(sizeof("testkey-")-1);
+    std::string master_key;
+    if (key == "1")       master_key = "012345678901234567890123456789012345";
```

Comment:
@yehudasa 
1) Would it be ok, if I just change it to "rgw-internal-testkey-1"? The actual security risk is only that someone will have idea to create on barbican key with such name.
2) Do we have some predefined macro define that we set in debug builds?